File: ui/src/commands/cheat.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { clamp } from 'es-toolkit';
import { invoke } from '@tauri-apps/api/core';

export function cheatSetBuildingLevel(coord: Coord, id: BuildingId, level: BuildingLevel) {
  level = Math.trunc(clamp(level, 0, 255));
  return invoke<null>('cheat_set_building_level', { coord, id, level });
}

export function cheatSetFood(food: number) {
  food = Math.trunc(Math.max(0, food));
  return invoke<null>('cheat_set_food', { food });
}

export function cheatSetIron(iron: number) {
  iron = Math.trunc(Math.max(0, iron));
  return invoke<null>('cheat_set_iron', { iron });
}

export function cheatSetMaxInfrastructure(coord: Coord) {
  return invoke<null>('cheat_set_max_infrastructure', { coord });
}

export function cheatSetMaxResources() {
  return invoke<null>('cheat_set_max_resources');
}

export function cheatSetMaxSiloResources() {
  return invoke<null>('cheat_set_max_silo_resources');
}

export function cheatSetMaxWarehouseResources() {
  return invoke<null>('cheat_set_max_warehouse_resources');
}

export function cheatSetResources(resources: Resources) {
  return invoke<null>('cheat_set_resources', { resources });
}

export function cheatSetStability(coord: Coord, stability: number) {
  stability = clamp(stability, 0.0, 1.0);
  return invoke<null>('cheat_set_stability', { coord, stability });
}

export function cheatSetStone(stone: number) {
  stone = Math.trunc(Math.max(0, stone));
  return invoke<null>('cheat_set_stone', { stone });
}

export function cheatSetWood(wood: number) {
  wood = Math.trunc(Math.max(0, wood));
  return invoke<null>('cheat_set_wood', { wood });
}

File: crates/nil-client/src/client/cheat.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use super::Client;
use crate::error::Result;
use nil_core::infrastructure::building::{BuildingId, BuildingLevel};
use nil_core::resource::{Food, Iron, Resources, Stone, Wood};
use nil_core::village::{Coord, Stability};

impl Client {
  /// POST `/cheat/infrastructure`
  pub async fn cheat_set_max_infrastructure(&self, coord: Coord) -> Result<()> {
    self
      .http
      .post("cheat/infrastructure", coord)
      .await
  }

  /// POST `/cheat/infrastructure/building`
  pub async fn cheat_set_building_level(
    &self,
    coord: Coord,
    id: BuildingId,
    level: BuildingLevel,
  ) -> Result<()> {
    self
      .http
      .post("cheat/infrastructure/building", (coord, id, level))
      .await
  }

  /// GET `/cheat/resources`
  pub async fn cheat_set_max_resources(&self) -> Result<()> {
    self.http.get("cheat/resources").await
  }

  /// POST `/cheat/resources`
  pub async fn cheat_set_resources(&self, resources: Resources) -> Result<()> {
    self
      .http
      .post("cheat/resources", resources)
      .await
  }

  /// POST `/cheat/resources/food`
  pub async fn cheat_set_food(&self, food: Food) -> Result<()> {
    self
      .http
      .post("cheat/resources/food", food)
      .await
  }

  /// POST `/cheat/resources/iron`
  pub async fn cheat_set_iron(&self, iron: Iron) -> Result<()> {
    self
      .http
      .post("cheat/resources/iron", iron)
      .await
  }

  /// GET `/cheat/resources/silo`
  pub async fn cheat_set_max_silo_resources(&self) -> Result<()> {
    self.http.get("cheat/resources/silo").await
  }

  /// POST `/cheat/resources/stone`
  pub async fn cheat_set_stone(&self, stone: Stone) -> Result<()> {
    self
      .http
      .post("cheat/resources/stone", stone)
      .await
  }

  /// GET `/cheat/resources/warehouse`
  pub async fn cheat_set_max_warehouse_resources(&self) -> Result<()> {
    self
      .http
      .get("cheat/resources/warehouse")
      .await
  }

  /// POST `/cheat/resources/wood`
  pub async fn cheat_set_wood(&self, wood: Wood) -> Result<()> {
    self
      .http
      .post("cheat/resources/wood", wood)
      .await
  }

  /// POST `/cheat/village/stability`
  pub async fn cheat_set_stability(&self, coord: Coord, stability: Stability) -> Result<()> {
    self
      .http
      .post("cheat/village/stability", (coord, stability))
      .await
  }
}
